src/main/java/nwoolcan/model/brewery/warehouse/WarehouseImpl.java

```diff
@@ -0,0 +1,92 @@
+package nwoolcan.model.brewery.warehouse;
+
+import nwoolcan.model.brewery.warehouse.article.Article;
+import nwoolcan.utils.Empty;
+import nwoolcan.utils.Result;
+
+import javax.annotation.Nullable;
+import java.util.Date;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Collection;
+import java.util.ArrayList;
+
+
+
+
+/**
+ * Warehouse implementation.
+ */
+public final class WarehouseImpl implements Warehouse {
+
+    private static final String ARTICLE_ALREADY_REGISTERED = "The article was already registered.";
+    private final Map<Stock, Stock> stocks;
+    private final Set<Article> articles;
+
+    /**
+     * Constructor of the Warehouse.
+     */
+    // Package protected
+    WarehouseImpl() {
+        this.stocks = new HashMap<>();
+        this.articles = new HashSet<>();
+    }
+
+    @Override
+    public Result<Collection<Stock>> getStocks(final QueryStock queryStock) {
+        return Result.of(new ArrayList<>()); //TODO once queryStock will be implemented.
+    }
+
+    @Override
+    public Result<Collection<Article>> getArticles(final QueryArticle queryArticle) {
+        return Result.of(new ArrayList<>()); //TODO once queryArticle will be implemented.
+    }
+
+    @Override
+    public Result<Empty> addArticle(final Article newArticle) {
+        return Result.ofEmpty()
+                     .require(() -> !this.articles.contains(newArticle), new IllegalArgumentException(ARTICLE_ALREADY_REGISTERED))
+                     .peek(res -> updateArticles(newArticle));
+    }
+
+    @Override
+    public Result<Empty> addRecord(final Article article, @Nullable final Date expirationDate, final Record record) {
+        return Result.of(new StockImpl(article, expirationDate))
+                     .peek(stock -> updateArticles(article))
+                     .flatMap(stock -> Result.of(this.getStock(stock)))
+                     .flatMap(stock -> stock.addRecord(record))
+                     .toEmpty();
+                     //.flatMap(stock -> Result.ofEmpty());
```

Comment:
useless

Reply:
it is not useless at all. the stock could change in that moment. be aware that the stock is hashed only by certain parameters.

Reply:
Please reply to the comments, don't change the comments. Thanks.